# All Campuses & Years - Data Wrangling
_**Campuses**: Madrid, Paris, Barcelona, Miami, Mexico, São Paulo, Berlin, Remote, Amsterdam and Lisbon.
**Years**: 2016-2021_

Data Wrangling is the process of cleaning, structuring and enriching raw data into a desired format for better decision making in less time.

- [Environment](#Environment)


- [Data Wrangling](#DataWrangling)
    - [Explore Data](#ExploreData)
    - [Clean Data](#CleanData)
    - [Remove Useless Columns](#RemoveColumns)
    - [Add New Columns](#AddColumns)


- [Clean CSV](#CleanCSV)


- [Main Insigths](#Insights)

## Environment

#### import libraries

In [1]:
# Pandas and Numpy
import pandas as pd
import numpy as np

#Time 
import time
from datetime import datetime
import datetime as dt

# Maths
import math
from scipy import stats

#Transform
from sklearn.preprocessing import OneHotEncoder

# Visualization
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

#### import data

In [2]:
apps = pd.read_csv("apps_ih_extended.csv", sep=";",low_memory=False) 

In [3]:
apps.tail()

Opportunity ID                                   Opportunity Name  \
40223  0065J000003zelq           Felix Funke - BER-WD-FT-(2020-10-19)-EUR   
40224  0065J000003zhAA          Marina Plana - BCN-UX-FT-(2020-09-28)-EUR   
40225  0065J000002KazQ                              jose zhu - DAPT - RMT   
40226  0065J000003ysu3       Andrea Gordillo - BER-DA-FT-(2020-10-19)-EUR   
40227  0065J000004J2Cx  Bárbara Guerbas - RMT-WD-FT-(2020-09-21)-BRL-A...   

       Has Courses Bootcamp Course Bootcamp Format  Bootcamp Year  \
40223            1              WD              FT         2020.0   
40224            1              UX              FT         2020.0   
40225            0             NaN             NaN            NaN   
40226            1              DA              FT         2020.0   
40227            1              WD              FT         2020.0   

      Bootcamp Start Date     Campus      Campus - Timezone Created Date  ...  \
40223          19/10/2020     Berlin                    BER   20/07/2020  ...   
40224          28/09/2020  Barcelona                    BCN   21/07/2020  ...   
40225                 NaN     Remote      RMT-Europe/Madrid    5/08/2020  ...   
40226          19/10/2020     Berlin                    BER    8/07/2020  ...   
40227          21/09/2020     Remote  RMT-America/Sao_Paulo   12/09/2020  ...   

       Age First touchpoint Date First Touchpoint type  \
40223  NaN      12/11/2020 14:20                  Call   
40224  NaN       3/09/2020 15:23                  Call   
40225  NaN                   NaN                   NaN   
40226  NaN                   NaN                   NaN   
40227  NaN                   NaN                   NaN   

      Hours Before First Touchpoint    Needs Financing  \
40223                       2762,34                NaN   
40224                       1071,10                NaN   
40225                           NaN                NaN   
40226                           NaN  Bildungsgutschein   
40227                           NaN                NaN   

      Person Account: City Before Course Person Account: Company Located in  \
40223                             Berlin                                NaN   
40224                                NaN                                NaN   
40225                                NaN                                NaN   
40226                             Bochum                                NaN   
40227                                NaN                                NaN   

      Person Account: Eligible to Work In Person Account: Linkedin URL  \
40223                                 Yes                          NaN   
40224                                 NaN                          NaN   
40225                                 NaN                          NaN   
40226                                 Yes                          NaN   
40227                                 NaN                          NaN   

      Programming languages  
40223                   NaN  
40224                   NaN  
40225                   NaN  
40226                   NaN  
40227                   NaN  

[5 rows x 58 columns]

## ExploreData

#### Metadata

- bootcamp_course: course associated to the opportunity, either WD, UX or DA.
- bootcamp_format: format associated to the opportunity, either FT or PT.
- bootcamp_start_date: Start Date of the course associated to thez opportunity.
- stage: The stage name the application lives at that moment.
- lost_deal_reason: Reason why the applicant has dropped the admission process
- stage_before_lost: The stage where the opportunity is before getting lost
- created_date: Date and time when this record was created. 
- close_date: Date when the opportunity is expected to close (before closing) or closed.
- drop_reason: Reason why the applicant has dropped after paying his deposit
- net_amount: Amount minus Discount or Scholarship Amount
- discount_type: Type of the discount, either Scholarship or Discount
- discount_name: Name of the discount initiative associated to the opportunity
- discount_amount: Amount of the discount that the applicant received
- scholarship_name: Name of the scholarship initiative associated to the opportunity
- scholarship_amount: Amount of the scholarship discount given to the opportunity
- financing_options: name of the financing option
- financing_option_amount: amount in local currency of the financing option
- deposit_payment: The amount of the tuition neither the student or Ironhack paid
- Stage Duration: The number of days the opportunity was in the stage listed in the Stage column
- ...

#### All variables info

In [4]:
apps.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40228 entries, 0 to 40227
Data columns (total 58 columns):
 #   Column                                    Non-Null Count  Dtype  
---  ------                                    --------------  -----  
 0   Opportunity ID                            40228 non-null  object 
 1   Opportunity Name                          40145 non-null  object 
 2   Has Courses                               40228 non-null  int64  
 3   Bootcamp Course                           36637 non-null  object 
 4   Bootcamp Format                           36635 non-null  object 
 5   Bootcamp Year                             36674 non-null  float64
 6   Bootcamp Start Date                       36674 non-null  object 
 7   Campus                                    39942 non-null  object 
 8   Campus - Timezone                         39942 non-null  object 
 9   Created Date                              40228 non-null  object 
 10  Close Date                        

In [5]:
apps.columns

Index(['Opportunity ID', 'Opportunity Name', 'Has Courses', 'Bootcamp Course',
       'Bootcamp Format', 'Bootcamp Year', 'Bootcamp Start Date', 'Campus',
       'Campus - Timezone', 'Created Date', 'Close Date', 'Invoice Date',
       'Stage', 'Discount Amount (converted) Currency',
       'Discount Amount (converted)',
       'Scholarship Amount (converted) Currency',
       'Scholarship Amount (converted)', 'Amount (converted) Currency',
       'Amount (converted)', 'Net Amount With VAT (converted) Currency',
       'Net Amount With VAT (converted)', 'Drop', 'Drop Reason',
       'Lost Deal Reason', 'Stage Before Lost', 'how did you hear about us?',
       'Person Account: Tech Background', 'Person Account: Gender',
       'Person Account: Highest degree received',
       'Person Account: Type of Position', 'Person Account: Looking for a Job',
       'Person Account: Nationality', 'Last Modified Date',
       'All Traffic Sources', 'First Click Medium', 'First Click Source',
       

In [6]:
apps.shape

(40228, 58)

In [7]:
apps.describe()

Has Courses  Bootcamp Year          Drop  Inactive Days  \
count  40228.000000   36674.000000  40228.000000    7315.000000   
mean       0.910386    2018.957109      0.010391      66.009296   
std        0.285632       3.016283      0.101405      53.712602   
min        0.000000    1900.000000      0.000000       0.000000   
25%        1.000000    2018.000000      0.000000      27.000000   
50%        1.000000    2019.000000      0.000000      56.000000   
75%        1.000000    2020.000000      0.000000      95.000000   
max        1.000000    2021.000000      1.000000     392.000000   

       Person Account: Company Located in  Programming languages  
count                                 0.0                    0.0  
mean                                  NaN                    NaN  
std                                   NaN                    NaN  
min                                   NaN                    NaN  
25%                                   NaN                    NaN  
50%                                   NaN                    NaN  
75%                                   NaN                    NaN  
max                                   NaN                    NaN

#### Numeric Variables

In [8]:
numeric= apps.select_dtypes(include=np.number)
numeric

Has Courses  Bootcamp Year  Drop  Inactive Days  \
0                0            NaN     0            NaN   
1                1         2018.0     0            NaN   
2                1         2019.0     0            NaN   
3                1         2020.0     0            NaN   
4                1         2018.0     0            NaN   
...            ...            ...   ...            ...   
40223            1         2020.0     0           25.0   
40224            1         2020.0     0           71.0   
40225            0            NaN     0            NaN   
40226            1         2020.0     0          153.0   
40227            1         2020.0     0           87.0   

       Person Account: Company Located in  Programming languages  
0                                     NaN                    NaN  
1                                     NaN                    NaN  
2                                     NaN                    NaN  
3                                     NaN                    NaN  
4                                     NaN                    NaN  
...                                   ...                    ...  
40223                                 NaN                    NaN  
40224                                 NaN                    NaN  
40225                                 NaN                    NaN  
40226                                 NaN                    NaN  
40227                                 NaN                    NaN  

[40228 rows x 6 columns]

#### Categorical Variables

In [9]:
categoric= apps.select_dtypes(exclude=np.number)
categoric

Opportunity ID                                   Opportunity Name  \
0      0065J000003yd1Z                           Aide Suarez - UXFT - MEX   
1      0061t000006yw67  José Alejandro Rivas - MEX-WD-FT-(2018-01-22)...   
2      0061t00000G0OVU         foteini chora - BER-UX-FT-(2019-10-14)-EUR   
3      0061t00000LplR7      Gaurav Gajavelli - MIA-DA-FT-(2020-06-08)-USD   
4      0061t000006yxLB          Maurice Goud - AMS-UX-FT-(2018-10-22)-EUR   
...                ...                                                ...   
40223  0065J000003zelq           Felix Funke - BER-WD-FT-(2020-10-19)-EUR   
40224  0065J000003zhAA          Marina Plana - BCN-UX-FT-(2020-09-28)-EUR   
40225  0065J000002KazQ                              jose zhu - DAPT - RMT   
40226  0065J000003ysu3       Andrea Gordillo - BER-DA-FT-(2020-10-19)-EUR   
40227  0065J000004J2Cx  Bárbara Guerbas - RMT-WD-FT-(2020-09-21)-BRL-A...   

      Bootcamp Course Bootcamp Format Bootcamp Start Date     Campus  \
0                 NaN             NaN                 NaN     Mexico   
1                  WD              FT          22/01/2018     Mexico   
2                  UX              FT          14/10/2019     Berlin   
3                  DA              FT           8/06/2020      Miami   
4                  UX              FT          22/10/2018  Amsterdam   
...               ...             ...                 ...        ...   
40223              WD              FT          19/10/2020     Berlin   
40224              UX              FT          28/09/2020  Barcelona   
40225             NaN             NaN                 NaN     Remote   
40226              DA              FT          19/10/2020     Berlin   
40227              WD              FT          21/09/2020     Remote   

           Campus - Timezone Created Date  Close Date Invoice Date  ...  \
0                        MEX    4/07/2020  30/09/2020          NaN  ...   
1                        MEX    9/11/2017  30/12/2017          NaN  ...   
2                        BER   29/08/2019   3/09/2019          NaN  ...   
3                        MIA    8/01/2020  22/01/2020          NaN  ...   
4                        AMS   17/07/2018  25/08/2018          NaN  ...   
...                      ...          ...         ...          ...  ...   
40223                    BER   20/07/2020  24/08/2020   25/08/2020  ...   
40224                    BCN   21/07/2020  30/09/2020          NaN  ...   
40225      RMT-Europe/Madrid    5/08/2020   5/08/2020          NaN  ...   
40226                    BER    8/07/2020  30/09/2020          NaN  ...   
40227  RMT-America/Sao_Paulo   12/09/2020  30/09/2020          NaN  ...   

      Account: Last Activity Account: Last Modified Date  Age  \
0                        NaN                   8/10/2020  NaN   
1                        NaN                  21/04/2020  NaN   
2                        NaN                  14/08/2020  NaN   
3                  8/01/2020                   3/09/2020  NaN   
4                        NaN                   6/05/2020  NaN   
...                      ...                         ...  ...   
40223             12/11/2020                  30/10/2020  NaN   
40224             28/09/2020                   2/10/2020  NaN   
40225              7/08/2020                   4/12/2020  NaN   
40226             12/08/2020                   9/10/2020  NaN   
40227             12/09/2020                  22/09/2020  NaN   

      First touchpoint Date First Touchpoint type  \
0                       NaN                   NaN   
1                       NaN                   NaN   
2                       NaN                   NaN   
3                       NaN                   NaN   
4                       NaN                   NaN   
...                     ...                   ...   
40223      12/11/2020 14:20                  Call   
40224       3/09/2020 15:23                  Call   
40225                   NaN                   NaN   
40226 

#### Nan values

In [10]:
apps.isnull().sum()

Opportunity ID                                  0
Opportunity Name                               83
Has Courses                                     0
Bootcamp Course                              3591
Bootcamp Format                              3593
Bootcamp Year                                3554
Bootcamp Start Date                          3554
Campus                                        286
Campus - Timezone                             286
Created Date                                    0
Close Date                                      0
Invoice Date                                37447
Stage                                           0
Discount Amount (converted) Currency        19739
Discount Amount (converted)                 19739
Scholarship Amount (converted) Currency     40118
Scholarship Amount (converted)              40118
Amount (converted) Currency                  3365
Amount (converted)                           3365
Net Amount With VAT (converted) Currency        0


In [11]:
total_values_xCol= apps.count()+apps.isnull().sum()

In [12]:
total_values= total_values_xCol.sum()

In [13]:
total_ExValues=apps.count().sum()

In [14]:
total_NanValues=apps.isnull().sum().sum()

In [15]:
print(round(total_NanValues/total_values*100,2),"% of our data is null")
print("*Taking into account all IH organic apps (40228) and the 58 columns from IH that refers to the user and its moves during the Admissions Process")


48.89 % of our data is null
*Taking into account all IH organic apps (40228) and the 58 columns from IH that refers to the user and its moves during the Admissions Process


## CleanData

#### Column: Opportunity Name

This column will not give us any value, as it is not important the name of the person who enrols. But, we will wait to remove it as it can be usefull to complete other columns with its information. Only Nans will be remove now as those are the ones that do not correspont to a real applicant.

In [16]:
apps.drop(apps[apps["Opportunity Name"].isnull()==True].index, inplace=True)                                 

#### Column: Has Courses

In [17]:
apps["Has Courses"].value_counts()

1    36541
0     3604
Name: Has Courses, dtype: int64

We will remove the application without course as it was from a test campaign that was on for a few month and paused from november 2020, and in this project we'd like to predict the random organic applicaiton.

Applicants without course correspond to 1% of our dataset

In [18]:
apps.drop(apps[apps["Has Courses"]==0].index, inplace=True)                                 

#### Column: Bootcamp Course

In [19]:
apps["Bootcamp Course"].isnull().sum()

27

In [20]:
apps["Bootcamp Course"].value_counts()

WD     21319
UX      8586
DA      6189
CY       401
ios       19
Name: Bootcamp Course, dtype: int64

In [21]:
apps[apps["Bootcamp Course"].isnull()==True]

Opportunity ID                                   Opportunity Name  \
2917   0061t000006yvz7                    Edith Orejel --(1970-01-01)-MXN   
4885   0061t000006yw1I                               0 --(1970-01-01)-EUR   
5592   0061t000006yuq2               borja monforte MAD--(2016-06-27)-EUR   
5658   0061t000006yurA        Eloy Fernández Gallut MAD--(2016-06-27)-EUR   
5887   0061t000006yw1V                               0 --(1970-01-01)-EUR   
5899   0061t000006yw1h                               0 --(1970-01-01)-EUR   
6671   0061t000006yyDH         Renato Tironi - SAO-DA-PT-(2019-03-19)-BRL   
6701   0061t000006yyDJ           Vitor Brito - SAO-DA-PT-(2019-03-19)-BRL   
8196   0061t000006yurL                 Juan Sanchez MAD--(2016-06-27)-EUR   
8226   0061t000006yurP              Izan Lopez Seco MAD--(2016-06-27)-EUR   
9564   0061t000006yuye     Stoilka Angelova Gyuleva MAD--(2016-09-12)-EUR   
10498  0061t000006yxBE   Matthias Staudinger - BER-WD-FT-(2018-10-08)-EUR   
14117  0061t000006yyDj       Miguel Queimado - SAO-DA-PT-(2019-03-19)-BRL   
14256  0061t000006yuqP          Jorge González Díaz MAD--(2016-09-12)-EUR   
15042  0061t000006yuqT                 Fabio Korsos MAD--(2016-06-27)-EUR   
15317  0061t000006yyCh  RAFAEL MAIA MARQUES MARQUES - SAO-WD-PT-(2019-...   
16895  0061t000006yyCk         Renato Tironi - SAO-DA-PT-(2019-03-19)-BRL   
17454  0061t000006yupt             Jacqueline Exton MAD--(2017-01-16)-EUR   
17466  0061t000006yur1        Servio Loayza Aguilar MAD--(2017-01-16)-EUR   
18191  0061t000006yyCE  José Francisco Pardo García MAD--(2019-04-22)-EUR   
18431  0061t000006yyCn          João  Bicudo - SAO-DA-PT-(2019-03-19)-BRL   
18743  0061t000006yyD2            Jaime Perez Lopez MAD--(2019-10-21)-EUR   
19926  0061t000006yyCX      Arturo García del Campo MAD--(2019-04-22)-EUR   
20891  0061t000006yupu               Ignacio Rivera MAD--(2016-06-27)-EUR   
22448  0061t000006yyD3               Alberto Varela MAD--(2019-04-22)-EUR   
22715  0061t000006yyDF        ANDRES RUIZ FERNANDEZ MAD--(2019-04-22)-EUR   
29237  0061t000006yyDK                 Pablo Deprit MAD--(2019-10-21)-EUR   

       Has Courses Bootcamp Course Bootcamp Format  Bootcamp Year  \
2917             1             NaN             NaN         1970.0   
4885             1             NaN             NaN         1970.0   
5592             1             NaN             NaN         2016.0   
5658             1             NaN             NaN         2016.0   
5887             1             NaN             NaN         1970.0   
5899             1             NaN             NaN         1970.0   
6671             1             NaN             NaN         2019.0   
6701             1             NaN             NaN         2019.0   
8196             1             NaN             NaN         2016.0   
8226             1             NaN             NaN         2016.0   
9564             1             NaN             NaN         2016.0   
10498            1             NaN             NaN         2018.0   
14117            1             NaN             NaN         2019.0   
14256            1             NaN             NaN         2016.0   
15042            1             NaN             NaN         2016.0   
15317            1             NaN             NaN         2019.0   
16895            1             NaN             NaN         2019.0   
17454            1             NaN             NaN         2017.0   
17466            1             NaN             NaN         2017.0   
18191            1             NaN             NaN         2019.0   
18431            1             NaN             NaN         2019.0   
18743            1             NaN             NaN         2019.0   
19926            1             NaN             NaN         2019.0   
20891            1             NaN             NaN         2016.0   
22448            1             NaN             NaN         2019.0   
22715            1             NaN             NaN       

As we see, applicants without Bootcamp Course lack of most of their information. Most of those applicants correspond to Madrid and Sao Paulo campus.

We decide to remove those 37 rows.

In [22]:
apps.drop(apps[apps["Bootcamp Course"].isnull()==True].index, inplace=True)                                 

#### Column: Bootcamp Format

In [23]:
apps["Bootcamp Format"].isnull().sum()

2

In [24]:
apps["Bootcamp Format"].value_counts()

FT    28054
PT     8458
Name: Bootcamp Format, dtype: int64

In [25]:
apps[apps["Bootcamp Format"].isnull()==True]

Opportunity ID                               Opportunity Name  \
2324   0061t00000JCVsu  Davide Bernocchi - BER-WD-FT-(2019-10-14)-EUR   
16507  0061t00000G0vzn  Alexandra Servie - BER-WD-FT-(2019-10-14)-EUR   

       Has Courses Bootcamp Course Bootcamp Format  Bootcamp Year  \
2324             1              WD             NaN         2019.0   
16507            1              WD             NaN         2019.0   

      Bootcamp Start Date  Campus Campus - Timezone Created Date  ...  Age  \
2324           14/10/2019  Berlin               BER   21/09/2019  ...  NaN   
16507          14/10/2019  Berlin               BER   11/09/2019  ...  NaN   

      First touchpoint Date First Touchpoint type  \
2324                    NaN                   NaN   
16507                   NaN                   NaN   

      Hours Before First Touchpoint    Needs Financing  \
2324                            NaN  Bildungsgutschein   
16507                           NaN                 No   

      Person Account: City Before Course Person Account: Company Located in  \
2324                              Berlin                                NaN   
16507                                NaN                                NaN   

      Person Account: Eligible to Work In Person Account: Linkedin URL  \
2324                                  Yes                          NaN   
16507                                 NaN                          NaN   

      Programming languages  
2324                    NaN  
16507                   NaN  

[2 rows x 58 columns]

In [26]:
apps["Bootcamp Format"].fillna("FT", inplace = True) 

#### Column: Bootcamp Year

In [27]:
apps["Bootcamp Year"].value_counts()

2020.0    13576
2019.0     9700
2018.0     6295
2021.0     2863
2017.0     2853
2016.0     1124
1970.0      101
1900.0        2
Name: Bootcamp Year, dtype: int64

In [28]:
apps["Bootcamp Year"].dtype

dtype('float64')

In [29]:
apps["Bootcamp Year"]=apps["Bootcamp Year"].astype(int)

In [30]:
apps[apps["Bootcamp Year"]==1970]

Opportunity ID                                   Opportunity Name  \
830    0061t000006yv9I        Andres Archila - MIA-WD-FT-(1970-01-01)-USD   
3678   0061t000006yvBX    Denise  Mcclintock - MIA-WD-FT-(1970-01-01)-USD   
3690   0061t000006yv8b           Mike Mullen - MIA-WD-FT-(1970-01-01)-USD   
4879   0061t000006yw0y                        0 MEX-WD-FT(1970-01-01)-EUR   
5352   0061t000006yv9N          Rachel Mahan - MIA-WD-FT-(1970-01-01)-USD   
...                ...                                                ...   
28310  0061t000006yv8N            Kat Alderman MIA-UX-FT-(1970-01-01)-USD   
38760  0061t000006yv9W  Andreza Feu Alvim Starling - MIA-WD-FT-(1970-0...   
39891  0061t000006yv99       Devorah Langsam - MIA-WD-FT-(1970-01-01)-USD   
39893  0061t000006yv8d          Greg Perlman - MIA-WD-FT-(1970-01-01)-USD   
39912  0061t000006yv9d          Mary Mederos - MIA-WD-FT-(1970-01-01)-USD   

       Has Courses Bootcamp Course Bootcamp Format  Bootcamp Year  \
830              1              WD              FT           1970   
3678             1              WD              FT           1970   
3690             1              WD              FT           1970   
4879             1              WD              FT           1970   
5352             1              WD              FT           1970   
...            ...             ...             ...            ...   
28310            1              UX              FT           1970   
38760            1              WD              FT           1970   
39891            1              WD              FT           1970   
39893            1              WD              FT           1970   
39912            1              WD              FT           1970   

      Bootcamp Start Date  Campus Campus - Timezone Created Date  ...  Age  \
830             1/01/1970   Miami               MIA    9/09/2016  ...  NaN   
3678            1/01/1970   Miami               MIA    9/09/2016  ...  NaN   
3690            1/01/1970   Miami               MIA    9/09/2016  ...  NaN   
4879            1/01/1970  Mexico               MEX   29/09/2017  ...  NaN   
5352            1/01/1970   Miami               MIA    9/09/2016  ...  NaN   
...                   ...     ...               ...          ...  ...  ...   
28310           1/01/1970   Miami               MIA    9/09/2016  ...  NaN   
38760           1/01/1970   Miami               MIA    9/09/2016  ...  NaN   
39891           1/01/1970   Miami               MIA    9/09/2016  ...  NaN   
39893           1/01/1970   Miami               MIA    9/09/2016  ...  NaN   
39912           1/01/1970   Miami               MIA    9/09/2016  ...  NaN   

      First touchpoint Date First Touchpoint type  \
830                     NaN                   NaN   
3678                    NaN                   NaN   
3690                    NaN                   NaN   
4879                    NaN                   NaN   
5352                    NaN                   NaN   
...                     ...                   ...   
28310                   NaN                   NaN   
38760                   NaN                   NaN   
39891                   NaN                   NaN   
39893                   NaN                   NaN   
39912                   NaN                   NaN   

      Hours Before First Touchpoint Needs Financing  \
830                             NaN             NaN   
3678                            NaN             NaN   
3690                            NaN             NaN   
4879                            NaN             NaN   
5352                            NaN             NaN   
...                             ...             ...   
28310                           NaN             NaN   
38760                           NaN             NaN   
39891                           NaN             NaN   
39893                           NaN             NaN   
39912                           NaN             NaN   

      Person Accoun

In [31]:
apps["Bootcamp Year"][apps["Bootcamp Year"]==1970]= apps["Created Date"].apply(lambda x: int(x[-4:]))
apps["Bootcamp Year"][apps["Bootcamp Year"]==1900]= apps["Created Date"].apply(lambda x: int(x[-4:]))

<ipython-input-31-d301c934db1f>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  apps["Bootcamp Year"][apps["Bootcamp Year"]==1970]= apps["Created Date"].apply(lambda x: int(x[-4:]))
<ipython-input-31-d301c934db1f>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  apps["Bootcamp Year"][apps["Bootcamp Year"]==1900]= apps["Created Date"].apply(lambda x: int(x[-4:]))


In [32]:
apps["Bootcamp Year"].value_counts()

2020    13577
2019     9701
2018     6295
2021     2863
2017     2856
2016     1222
Name: Bootcamp Year, dtype: int64

In [33]:
apps["Bootcamp Year"].isnull().sum()

0

#### Column: Bootcamp Start Date

In [34]:
apps["Bootcamp Start Date"].value_counts()

11/01/2021    1545
23/03/2020    1442
19/10/2020    1329
21/10/2019    1165
1/06/2020      833
              ... 
19/02/2019       1
17/07/2017       1
11/10/2016       1
3/04/2018        1
20/09/2021       1
Name: Bootcamp Start Date, Length: 298, dtype: int64

In [35]:
apps["Bootcamp Start Date"].isnull().sum()

0

In [36]:
apps["Bootcamp Start Date"]= pd.to_datetime(apps["Bootcamp Start Date"])

#### Column: Campus

In [37]:
apps["Campus"].value_counts()

Madrid       5856
Paris        5654
Barcelona    4920
Miami        4647
Mexico       3970
São Paulo    3089
Berlin       2801
Remote       2385
Amsterdam    1684
Lisbon       1361
Bogota        125
Munich         11
Name: Campus, dtype: int64

In [38]:
apps["Campus"].isnull().sum()

11

In [39]:
apps[apps["Campus"].isnull()==True]

Opportunity ID                                   Opportunity Name  \
413    0061t000009rosR           Marina Pereda MAD-UX-FT-(2019-06-03)-EUR   
4590   0061t000006yxiF         Lluis Arévalo Salom -WD-FT(2019-03-18)-EUR   
4614   0061t000006yxiM         Lluis Arévalo Salom -WD-FT(2019-03-11)-EUR   
4620   0061t000009s7Si    Lluis Arévalo Salom - MAD-WD-FT(2019-01-14)-EUR   
6997   0061t000006yw4e             Laure Magdelain -UX-FT(2018-06-11)-MXN   
11407  0061t000006yw4l                Manon Pellat -WD-FT(2018-01-22)-EUR   
15467  0061t000009s5ux  Hugues-Antoine Touchot de Vareilles - PAR-DA-F...   
21537  0061t000009sVQL           Khaled Badr - BER-UX-PT-(2019-06-04)-EUR   
28068  0061t000009s5jG         Aniruddh Jain - BER-WD-FT-(2019-06-03)-EUR   
28290  0061t000009s9Jv              Lluís Rayo BCN-DA-PT-(2019-05-28)-EUR   
28542  0061t000009sXIX           Jabel Riera - BCN-DA-PT-(2019-05-28)-EUR   

       Has Courses Bootcamp Course Bootcamp Format  Bootcamp Year  \
413              1              UX              FT           2019   
4590             1              WD              FT           2019   
4614             1              WD              FT           2019   
4620             1              WD              FT           2019   
6997             1              UX              FT           2018   
11407            1              WD              FT           2018   
15467            1              DA              FT           2019   
21537            1              UX              PT           2019   
28068            1              WD              FT           2019   
28290            1              DA              PT           2019   
28542            1              DA              PT           2019   

      Bootcamp Start Date Campus Campus - Timezone Created Date  ...  Age  \
413            2019-03-06    NaN               NaN   19/04/2019  ...  NaN   
4590           2019-03-19    NaN               NaN    4/10/2018  ...  NaN   
4614           2019-12-03    NaN               NaN    4/10/2018  ...  NaN   
4620           2019-01-15    NaN               NaN   29/04/2019  ...  NaN   
6997           2018-12-06    NaN               NaN   25/10/2017  ...  NaN   
11407          2018-01-23    NaN               NaN   27/10/2017  ...  NaN   
15467          2019-05-13    NaN               NaN   28/04/2019  ...  NaN   
21537          2019-04-06    NaN               NaN    7/05/2019  ...  NaN   
28068          2019-03-06    NaN               NaN   28/04/2019  ...  NaN   
28290          2019-05-28    NaN               NaN   29/04/2019  ...  NaN   
28542          2019-05-28    NaN               NaN    8/05/2019  ...  NaN   

      First touchpoint Date First Touchpoint type  \
413                     NaN                   NaN   
4590                    NaN                   NaN   
4614                    NaN                   NaN   
4620                    NaN                   NaN   
6997                    NaN                   NaN   
11407                   NaN                   NaN   
15467                   NaN                   NaN   
21537                   NaN                   NaN   
28068                   NaN                   NaN   
28290                   NaN                   NaN   
28542                   NaN                   NaN   

      Hours Before First Touchpoint Needs Financing  \
413                             NaN             Yes   
4590                            NaN             NaN   
4614                            NaN             NaN   
4620                            NaN             NaN   
6997                            NaN             NaN   
11407                           NaN             NaN   
15467                           NaN             NaN   
21537                           NaN             NaN   
28068                           NaN             NaN   
28290                           NaN             NaN   
28542                           NaN             NaN   

      Person Account: City Befo

we'll drop for now all applicants without campus as some are missing too much info and others are dupplicates.

In [40]:
apps.drop(apps[apps["Campus"].isnull()==True].index, inplace=True)                                 

Also, we'll drop applicants from Bogota & Munich since finally those campus were not open

In [41]:
apps.drop(apps[(apps["Campus"]=="Bogota") | (apps["Campus"]=="Munich")].index, inplace=True)        

In [42]:
apps["Campus"].value_counts()

Madrid       5856
Paris        5654
Barcelona    4920
Miami        4647
Mexico       3970
São Paulo    3089
Berlin       2801
Remote       2385
Amsterdam    1684
Lisbon       1361
Name: Campus, dtype: int64

#### Column: Campus - Timezone

In [43]:
apps["Campus - Timezone"].value_counts()

MAD                        5856
PAR                        5654
BCN                        4920
MIA                        4647
MEX                        3970
SAO                        3089
BER                        2801
RMT-Europe/Madrid          1795
AMS                        1684
LIS                        1361
RMT-America/Mexico_City     256
RMT-America/Sao_Paulo       163
RMT-America/Chicago         124
RMT-                         47
Name: Campus - Timezone, dtype: int64

*We undesrtand that RMT- is RMT-Europe/Madrid 

In [44]:
apps["Campus - Timezone"][apps["Campus - Timezone"]=="RMT-"]="RMT-Europe/Madrid"

<ipython-input-44-3fe39a5f7f1b>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  apps["Campus - Timezone"][apps["Campus - Timezone"]=="RMT-"]="RMT-Europe/Madrid"


In [45]:
apps["Campus - Timezone"][(apps["Campus - Timezone"]=="RMT-America/Chicago")|(apps["Campus - Timezone"]=="RMT-America/Sao_Paulo")| (apps["Campus - Timezone"]=="RMT-America/Mexico_City")]="RMT-America"


<ipython-input-45-654e0e32cc4e>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  apps["Campus - Timezone"][(apps["Campus - Timezone"]=="RMT-America/Chicago")|(apps["Campus - Timezone"]=="RMT-America/Sao_Paulo")| (apps["Campus - Timezone"]=="RMT-America/Mexico_City")]="RMT-America"


In [46]:
apps["Campus - Timezone"][apps["Campus - Timezone"]=="RMT-Europe/Madrid"]="RMT-Europe"

<ipython-input-46-9c07bd77bed1>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  apps["Campus - Timezone"][apps["Campus - Timezone"]=="RMT-Europe/Madrid"]="RMT-Europe"


In [47]:
apps["Campus - Timezone"].isnull().sum()

0

In [48]:
apps["Campus - Timezone"].value_counts()

MAD            5856
PAR            5654
BCN            4920
MIA            4647
MEX            3970
SAO            3089
BER            2801
RMT-Europe     1842
AMS            1684
LIS            1361
RMT-America     543
Name: Campus - Timezone, dtype: int64

#### Column: Created Date, Closed Date, Bootcamp Start Date and Last Modified Date

Fix same datetime format for all column dates

In [49]:
apps["Created Date"]= pd.to_datetime(apps["Created Date"])
apps["Close Date"]= pd.to_datetime(apps["Close Date"])
apps["Bootcamp Start Date"]= pd.to_datetime(apps["Bootcamp Start Date"])
apps["Invoice Date"]= pd.to_datetime(apps["Invoice Date"])
apps["Last Modified Date"]= pd.to_datetime(apps["Last Modified Date"])

Make sure all "Created dates" are < "Close Month", "Close Date".

In [50]:
print("Total rows with Close, Created Date, Invoice Date:")
apps[["Close Date","Created Date","Invoice Date"]].count()

Total rows with Close, Created Date, Invoice Date:


Close Date      36367
Created Date    36367
Invoice Date     2781
dtype: int64

In [51]:
print("Rows that have correct dates (Created Date<Close Date):")
apps[["Close Date","Created Date","Invoice Date"]][(apps["Created Date"]<apps["Close Date"])].count()

Rows that have correct dates (Created Date<Close Date):


Close Date      27206
Created Date    27206
Invoice Date     2042
dtype: int64

In [52]:
print("Rows that have wrong dates (Created Date>Close Date):")
apps[["Close Date","Created Date", "Invoice Date"]][(apps["Created Date"]>apps["Close Date"])].count()

Rows that have wrong dates (Created Date>Close Date):


Close Date      7570
Created Date    7570
Invoice Date     703
dtype: int64

In [53]:
print("Rows/Campus that have wrong dates (Created Date>Close Date):")

print("São Paulo:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Campus"]=="São Paulo")].count())
print("Paris:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Campus"]=="Paris")].count())
print("Barcelona:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Campus"]=="Barcelona")].count())
print("Miami:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Campus"]=="Miami")].count())
print("Mexico:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Campus"]=="Mexico")].count())
print("Madrid:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Campus"]=="Madrid")].count())
print("Berlin:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Campus"]=="Berlin")].count())
print("Amsterdam:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Campus"]=="Amsterdam")].count())
print("Remote:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Campus"]=="Remote")].count())
print("Lisbon:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Campus"]=="Lisbon")].count())



Rows/Campus that have wrong dates (Created Date>Close Date):
São Paulo: 481
Paris: 1303
Barcelona: 1038
Miami: 1157
Mexico: 877
Madrid: 1223
Berlin: 610
Amsterdam: 355
Remote: 288
Lisbon: 238


In [54]:
print("Rows/Stage that have wrong dates (Created Date>Close Date):")

print("Closed lost:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Stage"]=="Closed lost")].count())
print("Paid Deposit:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Stage"]=="Paid Deposit")].count())
print("Technical Interview:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Stage"]=="Technical Interview")].count())
print("Qualification:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Stage"]=="Qualification")].count())
print("Applied:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Stage"]=="Applied")].count())
print("Personal Interview:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Stage"]=="Personal Interview")].count())
print("Student Agreement:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Stage"]=="Student Agreement")].count())
print("ISA won on account:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Stage"]=="ISA won on account")].count())


Rows/Stage that have wrong dates (Created Date>Close Date):
Closed lost: 5513
Paid Deposit: 2028
Technical Interview: 2
Qualification: 6
Applied: 4
Personal Interview: 5
Student Agreement: 6
ISA won on account: 6


In [55]:
print("Rows/Format that have wrong dates (Created Date>Close Date):")

print("FT:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Bootcamp Format"]=="FT")].count())
print("PT:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Bootcamp Format"]=="PT")].count())

Rows/Format that have wrong dates (Created Date>Close Date):
FT: 5809
PT: 1761


In [56]:
print("Rows/Course that have wrong dates (Created Date>Close Date):")

print("WD:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Bootcamp Course"]=="WD")].count())
print("UX:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Bootcamp Course"]=="UX")].count())
print("CY:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Bootcamp Course"]=="CY")].count())
print("ios:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Bootcamp Course"]=="ios")].count())

Rows/Course that have wrong dates (Created Date>Close Date):
WD: 4610
UX: 1800
CY: 7
ios: 3


In [57]:
print("Rows/Year that have wrong dates (Created Date>Close Date):")

print("2021:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Bootcamp Year"]==2021)].count())
print("2020:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Bootcamp Year"]==2020)].count())
print("2019:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Bootcamp Year"]==2019)].count())
print("2018:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Bootcamp Year"]==2018)].count())
print("2017:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Bootcamp Year"]==2017)].count())
print("2016:",apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Bootcamp Year"]==2016)].count())

Rows/Year that have wrong dates (Created Date>Close Date):
2021: 178
2020: 2480
2019: 2348
2018: 1644
2017: 706
2016: 214


In [58]:
print("Total Rows/Year:")

print("2021:",apps["Campus"][(apps["Bootcamp Year"]==2021)].count())
print("2020:",apps["Campus"][(apps["Bootcamp Year"]==2020)].count())
print("2019:",apps["Campus"][(apps["Bootcamp Year"]==2019)].count())
print("2018:",apps["Campus"][(apps["Bootcamp Year"]==2018)].count())
print("2017:",apps["Campus"][(apps["Bootcamp Year"]==2017)].count())
print("2016:",apps["Campus"][(apps["Bootcamp Year"]==2016)].count())

Total Rows/Year:
2021: 2863
2020: 13577
2019: 9556
2018: 6293
2017: 2856
2016: 1222


In [59]:
y2021 =apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Bootcamp Year"]==2021)].count()
y2020=apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Bootcamp Year"]==2020)].count()
y2019=apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Bootcamp Year"]==2019)].count()
y2018=apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Bootcamp Year"]==2018)].count()
y2017=apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Bootcamp Year"]==2017)].count()
y2016=apps["Campus"][(apps["Created Date"]>apps["Close Date"])&(apps["Bootcamp Year"]==2016)].count()

In [60]:
ty2021 =apps["Campus"][(apps["Bootcamp Year"]==2021)].count()
ty2020=apps["Campus"][(apps["Bootcamp Year"]==2020)].count()
ty2019=apps["Campus"][(apps["Bootcamp Year"]==2019)].count()
ty2018=apps["Campus"][(apps["Bootcamp Year"]==2018)].count()
ty2017=apps["Campus"][(apps["Bootcamp Year"]==2017)].count()
ty2016=apps["Campus"][(apps["Bootcamp Year"]==2016)].count()

In [61]:
print("% of wrong Rows/Year:")

print("2021:",round(y2021/ty2021,2))
print("2020:",round(y2020/ty2020,2))
print("2019:",round(y2019/ty2019,2))
print("2018:",round(y2018/ty2018,2))
print("2017:",round(y2017/ty2017,2))
print("2016:",round(y2016/ty2016,2))

% of wrong Rows/Year:
2021: 0.06
2020: 0.18
2019: 0.25
2018: 0.26
2017: 0.25
2016: 0.18


As we could see, between 20%-30% of the data/year comes without a correct closed date. In order to decide if we could remove it or nor, we decided to check how balanced was the wrong data distributed to make sure that if we remove it it won't affect. And, we can see that among bootcamp types, formats, years and campus was balance, no wrong data was specificly comming from a place.


Nevertheless, since we have 706 invoice dates that could substitute the close dates, we will proceed to at least modifly this data and rise our target(y) data. This will help our model as so far our x and y are unbalanced.

In [62]:
apps["Close Date"]= np.where(apps["Invoice Date"].isnull()==False,apps["Invoice Date"], apps["Close Date"])

In [63]:
apps[["Close Date","Created Date","Invoice Date"]][(apps["Created Date"]<apps["Close Date"])].count()

Close Date      27179
Created Date    27179
Invoice Date     2015
dtype: int64

In [64]:
apps[["Close Date","Created Date", "Invoice Date"]][(apps["Created Date"]>apps["Close Date"])].count()

Close Date      7556
Created Date    7556
Invoice Date     689
dtype: int64

Finally we see that only 15 rows can be subtitute as only 15 invoice dates are < Created Date. Not much of a big add.

Now, we proceed by removeing those rows that still have a wrong Close Date

In [65]:
apps.drop(apps[apps["Created Date"]>apps["Close Date"]].index, inplace=True)

#### Column: Invoice Date

In [66]:
apps["Invoice Date"].isnull().sum()

26719

We check if all applicants without Invoice date is because they are not in the Paid Deposit Stage

In [67]:
apps[["Invoice Date","Stage","Close Date","Created Date"]][(apps["Invoice Date"].isnull()==True) & (apps["Stage"]=="Paid Deposit")]


Invoice Date         Stage Close Date Created Date
28             NaT  Paid Deposit 2019-06-29   2019-05-29
37             NaT  Paid Deposit 2017-12-13   2017-09-11
53             NaT  Paid Deposit 2018-01-24   2017-11-28
61             NaT  Paid Deposit 2017-04-17   2017-02-24
65             NaT  Paid Deposit 2019-02-04   2017-11-28
...            ...           ...        ...          ...
40137          NaT  Paid Deposit 2020-12-08   2020-07-17
40152          NaT  Paid Deposit 2020-07-27   2020-07-20
40168          NaT  Paid Deposit 2020-07-22   2020-07-20
40189          NaT  Paid Deposit 2020-07-22   2020-07-20
40198          NaT  Paid Deposit 2020-08-17   2020-07-20

[3067 rows x 4 columns]

In [68]:
apps[["Invoice Date","Stage","Close Date"]][(apps["Invoice Date"].isnull()==False) & (apps["Stage"]=="Paid Deposit")]


Invoice Date         Stage Close Date
26      2019-09-25  Paid Deposit 2019-09-25
99      2020-10-20  Paid Deposit 2020-10-20
105     2020-01-31  Paid Deposit 2020-01-31
116     2019-09-16  Paid Deposit 2019-09-16
117     2020-11-02  Paid Deposit 2020-11-02
...            ...           ...        ...
40173   2020-10-20  Paid Deposit 2020-10-20
40212   2020-10-20  Paid Deposit 2020-10-20
40221   2020-08-14  Paid Deposit 2020-08-14
40222   2020-10-27  Paid Deposit 2020-10-27
40223   2020-08-25  Paid Deposit 2020-08-25

[2088 rows x 3 columns]

2088 paid deposit have Invoice Date & 3067 paid deposit don't have Invoice Date. Although it helps us to see how poor is our data, in this project the Invoice date will have no use so we will later drop this column as it will be no longer usefull.

#### Column: Stage

In [69]:
apps["Stage"].value_counts()

Closed lost            22895
Paid Deposit            5155
Applied                  176
Personal Interview       151
Technical Interview      149
Qualification            134
Student Agreement        130
ISA won on account        21
Name: Stage, dtype: int64

#### Column: Stage Duration

In [70]:
apps["Stage Duration"].head()

1    1125,0000000000
4     875,0000000000
5     952,0000000000
6     147,0000000000
7    1125,0000000000
Name: Stage Duration, dtype: object

In [71]:
apps["Stage Duration"] = apps["Stage Duration"].str.replace(",",".")

In [72]:
apps["Stage Duration"].fillna(0, inplace = True) 

In [73]:
apps["Stage Duration"] = apps["Stage Duration"].apply(pd.to_numeric, errors='coerce')

#### Column: Discount Amount (converted)

In [74]:
apps["Discount Amount (converted)"].isnull().sum()

13473

In [75]:
apps["Discount Amount (converted)"].head()

1      0,00
4      0,00
5      0,00
6       NaN
7    884,96
Name: Discount Amount (converted), dtype: object

In [76]:
apps["Discount Amount (converted)"] = apps["Discount Amount (converted)"].str.replace(",",".")

In [77]:
apps["Discount Amount (converted)"].fillna(0, inplace = True) 

In [78]:
apps["Discount Amount (converted)"] = apps["Discount Amount (converted)"].apply(pd.to_numeric, errors='coerce')

#### Column: Drop Reason

In [79]:
apps["Drop Reason"].isnull().sum()

28532

In [80]:
apps["Drop Reason"].value_counts()

Personal matters              92
COVID-19                      83
Can not afford the payment    26
Not interested                18
Changed my mind               14
Got a job                     13
No time                       13
Bad candidate                 11
Missed in application          4
Joined another bootcamp        3
No language skills             2
Name: Drop Reason, dtype: int64

In [81]:
apps["Drop Reason"].fillna("Not specified", inplace = True) 

#### Column: Amount (converted) 

In [82]:
apps["Amount (converted)"].isnull().sum()

0

In [83]:
apps["Amount (converted)"] = apps["Amount (converted)"].str.replace(",",".")

apps["Amount (converted)"] = apps["Amount (converted)"].apply(pd.to_numeric, errors='coerce')

In [84]:
apps["Amount (converted)"].value_counts()

7500.00     6688
6500.00     4799
7000.00     4359
8000.00     2525
3225.81     1815
4000.00     1792
9734.51     1039
4147.47      831
11061.95     818
10176.99     790
3764.71      771
6000.00      706
10619.47     623
4746.54      381
11504.42     301
3539.82      214
4352.94      120
8500.00       58
4588.24       54
61946.90      29
70000.00      17
3741.94       14
16000.00      12
14000.00       6
15000.00       5
15500.00       4
4705.88        4
6637.17        4
11000.00       3
5752.21        3
0.92           3
11500.00       2
21681.42       2
8407.08        2
14500.00       2
5100.00        1
2800.00        1
2705.88        1
23008.85       1
322.58         1
13000.00       1
8849.56        1
12500.00       1
22123.89       1
9000.00        1
5000.00        1
21238.94       1
8705.88        1
22566.37       1
20353.98       1
Name: Amount (converted), dtype: int64

13,000 USD (10,740 EUR) should be the highest tuition fee, as Miami part time courses are have the highest fee.
So we'll check that the max Amount is 10,740 EUR. We'll goe for 12000 as USD exchanged rate have changed.

In [85]:
apps[apps["Amount (converted)"]>12000]

Opportunity ID                                   Opportunity Name  \
823    0061t000006yvwo  Víctor Manuel Horta Osorio MEX-WD-FT-(2018-01-...   
851    0061t000006yxMO  Danielle De avila da silva - SAO-WD-FT-(2018-0...   
925    0061t000006yw3Z        Daniel Alarcon - MEX-WD-PT-(2018-04-24)-MXN   
1560   0065J000003HbIy          GABRIEL ZUNE - MIA-UX-FT-(2020-08-17)-USD   
1669   0061t000006yw9u       Marilyn Negrete - MEX-WD-FT-(2018-01-22)-MXN   
...                ...                                                ...   
37310  0065J000006CEC1        Julien FURNARI - PAR-WD-PT-(2021-01-12)-EUR   
38125  0065J000002D1Fl    Alessandro Broggio - BER-DA-PT-(2020-08-25)-EUR   
39208  0065J000003HxUI    Charlotte Pedersen - PAR-UX-FT-(2020-11-09)-EUR   
39745  0065J000002JdfC      Lilit Arutyunyan - LIS-DA-PT-(2020-09-15)-EUR   
39782  0065J000002CR22     Enrique Cerrolaza - MAD-DA-FT-(2020-10-19)-EUR   

       Has Courses Bootcamp Course Bootcamp Format  Bootcamp Year  \
823              1              WD              FT           2018   
851              1              WD              FT           2018   
925              1              WD              PT           2018   
1560             1              UX              FT           2020   
1669             1              WD              FT           2018   
...            ...             ...             ...            ...   
37310            1              WD              PT           2021   
38125            1              DA              PT           2020   
39208            1              UX              FT           2020   
39745            1              DA              PT           2020   
39782            1              DA              FT           2020   

      Bootcamp Start Date     Campus Campus - Timezone Created Date  ...  Age  \
823            2018-01-22     Mexico               MEX   2017-08-29  ...  NaN   
851            2018-08-20  São Paulo               SAO   2018-07-21  ...  NaN   
925            2018-04-23     Mexico               MEX   2017-10-16  ...  NaN   
1560           2020-08-17      Miami               MIA   2020-04-30  ...  NaN   
1669           2018-01-22     Mexico               MEX   2017-11-27  ...  NaN   
...                   ...        ...               ...          ...  ...  ...   
37310          2021-02-16      Paris               PAR   2020-11-25  ...  NaN   
38125          2020-08-25     Berlin               BER   2020-06-25  ...  NaN   
39208          2020-09-11      Paris               PAR   2020-06-05  ...  NaN   
39745          2020-10-20     Lisbon               LIS   2020-04-06  ...  NaN   
39782          2020-10-19     Madrid               MAD   2020-06-13  ...  NaN   

      First touchpoint Date First Touchpoint type  \
823                     NaN                   NaN   
851                     NaN                   NaN   
925                     NaN                   NaN   
1560                    NaN                   NaN   
1669                    NaN                   NaN   
...                     ...                   ...   
37310      26/11/2020 16:31                  Call   
38125      25/06/2020 13:48                  Call   
39208       6/05/2020 16:35                  Call   
39745       7/07/2020 19:19                  Call   
39782      15/06/2020 17:04                  Call   

      Hours Before First Touchpoint  Needs Financing  \
823                             NaN              NaN   
851                             NaN              NaN   
925                             NaN              NaN   
1560                            NaN              NaN   
1669                            NaN              NaN   
...                             ...              ...   
37310                         22,60              NaN   
38125                          0,07              Yes   
39208                          3,21              NaN   
39745                        800,50              NaN   
39782                       

We've seen that 84 rows have a wrong Amount and we don't know the reason although it could be that it doubles the real amount or move the comma 1 position, among other possible reasons. Therefore, for now we'll drop them.

In [86]:
apps.drop(apps[apps["Amount (converted)"]>12000].index, inplace=True)        

Same check we'll do for low amount. The lower tuiton fee goes to Brazil that is 18500 Brazilian Real (3012,98 EUR). We check it with 2500 as before BRazil had a lower fee.

In [87]:
apps[apps["Amount (converted)"]<2500]

Opportunity ID                                Opportunity Name  \
10379  0061t000006yxgy        Ines Guerra - MAD-WD-FT-(2019-01-07)-MXN   
12496  0061t000006yxJg  JuanCarlos Cortez - MEX-UX-PT-(2021-05-04)-MXN   
33623  0065J000004mRQL   Jonathan Ramirez - MEX-UX-PT-(2021-05-04)-MXN   
34675  0065J000004mSvj   Jesualdo jsualdo - MEX-UX-PT-(2021-05-04)-MXN   

       Has Courses Bootcamp Course Bootcamp Format  Bootcamp Year  \
10379            1              WD              FT           2019   
12496            1              UX              PT           2021   
33623            1              UX              PT           2021   
34675            1              UX              PT           2021   

      Bootcamp Start Date  Campus Campus - Timezone Created Date  ...  Age  \
10379          2019-07-01  Madrid               MAD   2018-01-10  ...  NaN   
12496          2021-05-04  Mexico               MEX   2018-07-13  ...  NaN   
33623          2021-05-04  Mexico               MEX   2020-02-10  ...  NaN   
34675          2021-05-04  Mexico               MEX   2020-03-10  ...  NaN   

      First touchpoint Date First Touchpoint type  \
10379                   NaN                   NaN   
12496       4/11/2020 22:44                  Call   
33623                   NaN                   NaN   
34675                   NaN                   NaN   

      Hours Before First Touchpoint  Needs Financing  \
10379                           NaN              NaN   
12496                      20297,69              NaN   
33623                           NaN              NaN   
34675                           NaN              NaN   

      Person Account: City Before Course Person Account: Company Located in  \
10379                                NaN                                NaN   
12496                               cdmx                                NaN   
33623                                NaN                                NaN   
34675                                NaN                                NaN   

      Person Account: Eligible to Work In  Person Account: Linkedin URL  \
10379                                 NaN                           NaN   
12496                                 Yes                           NaN   
33623                                 NaN                           NaN   
34675                                 NaN                           NaN   

      Programming languages  
10379                   NaN  
12496                   NaN  
33623                   NaN  
34675                   NaN  

[4 rows x 58 columns]

In [88]:
apps.drop(apps[apps["Amount (converted)"]<2500].index, inplace=True)        

#### Column: Net Amount With VAT (converted) 

In [89]:
apps["Net Amount With VAT (converted)"].isnull().sum()

0

In [90]:
apps["Net Amount With VAT (converted)"] = apps["Net Amount With VAT (converted)"].str.replace(",",".")

In [91]:
apps["Net Amount With VAT (converted)"] = apps["Net Amount With VAT (converted)"].apply(pd.to_numeric, errors='coerce')

#### Column: how did you hear about us?

In [92]:
apps["how did you hear about us?"].isnull().sum()

1096

In [93]:
apps["how did you hear about us?"].value_counts()

google                       6697
facebook                     4455
other                        3469
instagram                    1776
friendOrFamily               1647
Friend or family             1442
linkedin                      980
friend_or_family              831
youtube                       773
ironhackAlumni                736
switchup                      653
Ironhack alumni               543
newspaperOrMagazine           430
Newspaper/magazine            355
Ironhack event                346
ironhackEvent                 292
course_report                 282
ironhack_alumn                208
quora                         183
newspaper_or_magazine         173
ironhack_event                157
Blog                          141
courseReport                  140
Twitter                       132
Application form              111
ironhack_alum                  97
Social Media                   87
Workshop                       76
We/Code                        67
pole_emploi   

In [94]:
apps["how did you hear about us?"].fillna("Not specified", inplace = True) 

We can see that same reasons have diferent names, so below we homogenize them.

In [95]:
apps["how did you hear about us?"][(apps["how did you hear about us?"]=="Friend or family")
                                    |(apps["how did you hear about us?"]=="friend_or_family")
                                    |(apps["how did you hear about us?"]=="friendOrFamily")
                                   |(apps["how did you hear about us?"]=="Ironhack alumni")
                                    |(apps["how did you hear about us?"]=="ironhack_alumn")
                                    |(apps["how did you hear about us?"]=="ironhack_alum")
                                   |(apps["how did you hear about us?"]=="ironhackAlumni"
                                    )]="Referral"

apps["how did you hear about us?"][(apps["how did you hear about us?"]=="youtube")
                                    |(apps["how did you hear about us?"]=="linkedin")
                                    |(apps["how did you hear about us?"]=="friendOrFamily")
                                   |(apps["how did you hear about us?"]=="Twitter")
                                    |(apps["how did you hear about us?"]=="facebook")
                                    |(apps["how did you hear about us?"]=="instagram"
                                    )]="Social Media"

apps["how did you hear about us?"][(apps["how did you hear about us?"]=="S-Bahn"
                                    )]="sbahn" 
                                    
apps["how did you hear about us?"][(apps["how did you hear about us?"]=="Switch Up"
                                    )]="switchup" 
                                    
apps["how did you hear about us?"][(apps["how did you hear about us?"]=="course_report")
                                    |(apps["how did you hear about us?"]=="Course Report"
                                    )]="courseReport"                                
                                    
apps["how did you hear about us?"][(apps["how did you hear about us?"]=="Newspaper/magazine"
                                    )]="newspaperOrMagazine"  
                                    
apps["how did you hear about us?"][(apps["how did you hear about us?"]=="pole_emploi"
                                   )|(apps["how did you hear about us?"]=="Pole Emploi"
                                    )]="PoleEmploi"   
                                    
apps["how did you hear about us?"][(apps["how did you hear about us?"]=="Ironhack event")
                                    |(apps["how did you hear about us?"]=="ironhack_event")
                                    |(apps["how did you hear about us?"]=="ironhackEvent")
                                   |(apps["how did you hear about us?"]=="circus")
                                   |(apps["how did you hear about us?"]=="Paris Code")
                                   |(apps["how did you hear about us?"]=="Hackshow")
                                   |(apps["how did you hear about us?"]=="paris_code")
                                   |(apps["how did you hear about us?"]=="OpenHouse")
                                   |(apps["how did you hear about us?"]=="Open House")
                                   |(apps["how did you hear about us?"]=="Register to our OpenHouse")
                                   |(apps["how did you hear about us?"]=="We/Code")
                                   |(apps["how did you hear about us?"]=="Workshop"
                                   )]="IronhackEvent"                                                                            
                                    
apps["how did you hear about us?"][(apps["how did you hear about us?"]=="metro_bus"
                                   )]="Metro/Bus" 

<ipython-input-95-d3e598d679c7>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  apps["how did you hear about us?"][(apps["how did you hear about us?"]=="Friend or family")
<ipython-input-95-d3e598d679c7>:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  apps["how did you hear about us?"][(apps["how did you hear about us?"]=="youtube")
<ipython-input-95-d3e598d679c7>:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  apps["how did you h

In [96]:
apps["how did you hear about us?"].value_counts()

Social Media              8203
google                    6697
Referral                  5504
other                     3469
Not specified             1096
IronhackEvent             1045
newspaperOrMagazine        785
switchup                   654
courseReport               424
quora                      183
newspaper_or_magazine      173
Blog                       141
Application form           111
PoleEmploi                  94
sbahn                       25
TV and/or radio             25
IronStar                    18
Syllabus form               16
Challenge                   15
Pompeii Scholarship         14
Metro/Bus                    7
Ironhack                     4
ISA                          3
Article about Ironhack       3
Skillied                     3
Newsletter                   3
Selina                       2
salon                        2
Contact form                 1
Poster                       1
Meetup                       1
Hemper                       1
Name: ho

#### Column: Lost Deal Reason

In [97]:
apps["Lost Deal Reason"].value_counts()

Missed in application           6519
Not interested                  5433
Can not afford the payment      2615
Bad candidate                   1616
No time                         1000
Just wanted information          992
Personal matters                 729
Changed my mind                  714
Duplicated                       513
Next year                        508
Joined another bootcamp          500
Test application                 425
Got a job                        355
Fake application                 290
No language skills               204
Applied for Scholarship          142
COVID-19                         128
Looking for an online option     123
Different TimeZone                86
Unexpected course price           77
No English skills                 41
No Spanish skills                  3
Name: Lost Deal Reason, dtype: int64

In [98]:
apps["Lost Deal Reason"][(apps["Lost Deal Reason"]=="No English skills"
                                   )|(apps["how did you hear about us?"]=="No Spanish skills"
                                    )]="No language skills"  

<ipython-input-98-97a55615068f>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  apps["Lost Deal Reason"][(apps["Lost Deal Reason"]=="No English skills"


In [99]:
apps[(apps["Lost Deal Reason"].isnull()==True) & (apps["Stage"]=="Closed Lost") ]

Empty DataFrame
Columns: [Opportunity ID, Opportunity Name, Has Courses, Bootcamp Course, Bootcamp Format, Bootcamp Year, Bootcamp Start Date, Campus, Campus - Timezone, Created Date, Close Date, Invoice Date, Stage, Discount Amount (converted) Currency, Discount Amount (converted), Scholarship Amount (converted) Currency, Scholarship Amount (converted), Amount (converted) Currency, Amount (converted), Net Amount With VAT (converted) Currency, Net Amount With VAT (converted), Drop, Drop Reason, Lost Deal Reason, Stage Before Lost, how did you hear about us?, Person Account: Tech Background, Person Account: Gender, Person Account: Highest degree received, Person Account: Type of Position, Person Account: Looking for a Job, Person Account: Nationality, Last Modified Date, All Traffic Sources, First Click Medium, First Click Source, Last Click Medium, Last Click Source, Stage Duration, Inactive Days, Last Activity Date, Last Call Date, Last Event Date, Last Task Date, Last Email Date, Account: Created Date, Account: Last Activity, Account: Last Modified Date, Age, First touchpoint Date, First Touchpoint type, Hours Before First Touchpoint, Needs Financing, Person Account: City Before Course, Person Account: Company Located in, Person Account: Eligible to Work In, Person Account: Linkedin URL, Programming languages]
Index: []

[0 rows x 58 columns]

We can see that all applicants without Lost Deal Reason is because they are still open. So we will replace Nan for "open applicant"

In [100]:
apps["Lost Deal Reason"].fillna("Open applicant", inplace = True) 

#### Column: Scholarship Amount (converted)

In [101]:
apps["Scholarship Amount (converted)"] = apps["Scholarship Amount (converted)"].str.replace(",",".")

In [102]:
apps["Scholarship Amount (converted)"].fillna(0, inplace = True) 

In [103]:
apps["Scholarship Amount (converted)"] = apps["Scholarship Amount (converted)"].apply(pd.to_numeric, errors='coerce')

#### Column: Stage Before Lost

In [104]:
apps["Stage Before Lost"].value_counts()

Personal Interview     7748
Applied                5536
Technical Interview    3685
Qualification          3597
Student Agreement      1671
Paid Deposit             65
Video Interview          13
Scheduled Interview       1
ISA won on account        1
Name: Stage Before Lost, dtype: int64

In [105]:
apps[(apps["Stage Before Lost"].isnull()==True) & (apps["Stage"]=="Closed Lost") ]

Empty DataFrame
Columns: [Opportunity ID, Opportunity Name, Has Courses, Bootcamp Course, Bootcamp Format, Bootcamp Year, Bootcamp Start Date, Campus, Campus - Timezone, Created Date, Close Date, Invoice Date, Stage, Discount Amount (converted) Currency, Discount Amount (converted), Scholarship Amount (converted) Currency, Scholarship Amount (converted), Amount (converted) Currency, Amount (converted), Net Amount With VAT (converted) Currency, Net Amount With VAT (converted), Drop, Drop Reason, Lost Deal Reason, Stage Before Lost, how did you hear about us?, Person Account: Tech Background, Person Account: Gender, Person Account: Highest degree received, Person Account: Type of Position, Person Account: Looking for a Job, Person Account: Nationality, Last Modified Date, All Traffic Sources, First Click Medium, First Click Source, Last Click Medium, Last Click Source, Stage Duration, Inactive Days, Last Activity Date, Last Call Date, Last Event Date, Last Task Date, Last Email Date, Account: Created Date, Account: Last Activity, Account: Last Modified Date, Age, First touchpoint Date, First Touchpoint type, Hours Before First Touchpoint, Needs Financing, Person Account: City Before Course, Person Account: Company Located in, Person Account: Eligible to Work In, Person Account: Linkedin URL, Programming languages]
Index: []

[0 rows x 58 columns]

We can see that all applicants without Lost Deal Reason is because they are still open. So we will replace Nan for "open applicant"

In [106]:
apps["Stage Before Lost"].fillna("Open applicant", inplace = True) 

#### Column: Person Account: Gender

In [107]:
apps["Person Account: Gender"].value_counts()

Male                      17599
Female                     9728
M                           373
F                           190
Prefer not to disclose       79
Non-binary                   75
Hombre                       63
Mujer                        47
Otro                          2
Name: Person Account: Gender, dtype: int64

In [108]:
apps["Person Account: Gender"][(apps["Person Account: Gender"]=="M") | (apps["Person Account: Gender"]=="Hombre")] = "Male"
apps["Person Account: Gender"][(apps["Person Account: Gender"]=="F") | (apps["Person Account: Gender"]=="Mujer")] = "Female"
apps["Person Account: Gender"][(apps["Person Account: Gender"]=="Otro") | (apps["Person Account: Gender"]=="Mujer")] = "Prefer not to disclose"


<ipython-input-108-7001505b9d7a>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  apps["Person Account: Gender"][(apps["Person Account: Gender"]=="M") | (apps["Person Account: Gender"]=="Hombre")] = "Male"
<ipython-input-108-7001505b9d7a>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  apps["Person Account: Gender"][(apps["Person Account: Gender"]=="F") | (apps["Person Account: Gender"]=="Mujer")] = "Female"
<ipython-input-108-7001505b9d7a>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/us

In [109]:
apps["Person Account: Gender"].fillna("Prefer not to disclose", inplace = True) 

#### Columns: All other columns

In [110]:
apps.isnull().sum()

Opportunity ID                                  0
Opportunity Name                                0
Has Courses                                     0
Bootcamp Course                                 0
Bootcamp Format                                 0
Bootcamp Year                                   0
Bootcamp Start Date                             0
Campus                                          0
Campus - Timezone                               0
Created Date                                    0
Close Date                                      0
Invoice Date                                26633
Stage                                           0
Discount Amount (converted) Currency        13444
Discount Amount (converted)                     0
Scholarship Amount (converted) Currency     28637
Scholarship Amount (converted)                  0
Amount (converted) Currency                     0
Amount (converted)                              0
Net Amount With VAT (converted) Currency        0


We won't clean for now the rest of the columns as most of it are null values.

#### Save semi clean & clean data to csv

In [111]:
apps.to_csv("apps_allYears_semiClean.csv", index=False)

## RemoveColumns 
To continue, we remove not usefull columns and those that have +50% of null data (data we couldn't clean as missing information), to have a 100% clean df to transform and add new columns

In [112]:
apps.drop(columns=["Opportunity ID", "Opportunity Name", "Has Courses", 
                   "Campus","Invoice Date","Hours Before First Touchpoint",
                   "Person Account: Looking for a Job","Person Account: Tech Background",
                  "Person Account: Highest degree received","Person Account: Nationality","All Traffic Sources",
                   "First Click Medium","First Click Source", "Last Click Medium", "Last Click Source",
                   "Last Modified Date","Last Activity Date", "Last Call Date", "Last Event Date", 
                   "Last Task Date", "Last Email Date", "Account: Created Date", "Account: Last Activity",
                   "Account: Last Modified Date", "Age", "First touchpoint Date", "First Touchpoint type",
                   "Hours Before First Touchpoint","Needs Financing", "Person Account: City Before Course", 
                   "Person Account: Company Located in","Person Account: Eligible to Work In", 
                   "Person Account: Linkedin URL", "Programming languages", "Inactive Days",
                   "Discount Amount (converted) Currency", "Scholarship Amount (converted) Currency", 
                   "Amount (converted) Currency","Net Amount With VAT (converted) Currency",
                   "Lost Deal Reason","Stage Before Lost", "Person Account: Type of Position"
                  ], inplace=True)

In [113]:
apps.head()

Bootcamp Course Bootcamp Format  Bootcamp Year Bootcamp Start Date  \
1              WD              FT           2018          2018-01-22   
4              UX              FT           2018          2018-10-22   
5              UX              FT           2018          2018-08-20   
6              WD              FT           2020          2020-08-17   
7              WD              PT           2018          2018-11-06   

  Campus - Timezone Created Date Close Date        Stage  \
1               MEX   2017-09-11 2017-12-30  Closed lost   
4               AMS   2018-07-17 2018-08-25  Closed lost   
5               BCN   2018-01-05 2018-05-28  Closed lost   
6               MIA   2020-07-14 2020-07-14  Closed lost   
7               MIA   2017-09-11 2018-10-04  Closed lost   

   Discount Amount (converted)  Scholarship Amount (converted)  \
1                         0.00                             0.0   
4                         0.00                             0.0   
5                         0.00                             0.0   
6                         0.00                             0.0   
7                       884.96                             0.0   

   Amount (converted)  Net Amount With VAT (converted)  Drop    Drop Reason  \
1             3225.81                          3225.81     0  Not specified   
4             6000.00                          6000.00     0  Not specified   
5             6500.00                          6500.00     0  Not specified   
6            11061.95                         11061.95     0  Not specified   
7            10176.99                          9292.04     0  Not specified   

  how did you hear about us? Person Account: Gender  Stage Duration  
1               Social Media                   Male          1125.0  
4                     google                   Male           875.0  
5                     google                   Male           952.0  
6               Social Media                 Female           147.0  
7                      other                   Male          1125.0

In [114]:
apps.shape

(28723, 17)

In [115]:
apps.to_csv("apps_allYears_clean_selCols.csv", index=False)

## AddColumns

In [116]:
apps_clean = pd.read_csv("apps_allYears_clean_selCols.csv", parse_dates=["Bootcamp Start Date","Created Date","Close Date"])

#### Add Column: Paid Deposit 
This column will be our target

In [117]:
apps_clean["Paid Deposit"]= [1 if (value == "Paid Deposit")| (value == "ISA won on account")else 0 for value in apps_clean["Stage"]]

#### Add Column: Scholarship & Discount

In [118]:
apps_clean["Scholarship"]= [0 if value == 0.00 else 1 for value in apps_clean["Scholarship Amount (converted)"]]
apps_clean["Discount"]= [0 if value == 0.00 else 1 for value in apps_clean["Discount Amount (converted)"]]

#### Add Column: Time Conversion (days)
Time taken to convert the lead from it's creation to it's closure

In [119]:
apps_clean["Time Conversion - days"]= (apps_clean['Close Date'] - apps_clean['Created Date']).dt.days

In [120]:
apps_clean[(apps_clean['Close Date'] < apps_clean['Created Date']) &(apps_clean["Bootcamp Year"]==2020)].head()

Empty DataFrame
Columns: [Bootcamp Course, Bootcamp Format, Bootcamp Year, Bootcamp Start Date, Campus - Timezone, Created Date, Close Date, Stage, Discount Amount (converted), Scholarship Amount (converted), Amount (converted), Net Amount With VAT (converted), Drop, Drop Reason, how did you hear about us?, Person Account: Gender, Stage Duration, Paid Deposit, Scholarship, Discount, Time Conversion - days]
Index: []

[0 rows x 21 columns]

In [121]:
apps_clean[(apps_clean['Close Date'] > apps_clean['Created Date']) &(apps_clean["Bootcamp Year"]==2020)].head()

Bootcamp Course Bootcamp Format  Bootcamp Year Bootcamp Start Date  \
11              WD              FT           2020          2020-08-17   
19              WD              FT           2020          2020-09-14   
24              WD              FT           2020          2020-09-14   
26              UX              PT           2020          2020-08-01   
33              WD              FT           2020          2020-03-23   

   Campus - Timezone Created Date Close Date        Stage  \
11               MIA   2020-07-08 2020-09-30  Closed lost   
19        RMT-Europe   2020-03-09 2020-09-30  Closed lost   
24        RMT-Europe   2020-08-16 2020-09-30  Closed lost   
26               SAO   2019-08-29 2019-09-24  Closed lost   
33               MAD   2019-08-29 2020-02-17  Closed lost   

    Discount Amount (converted)  Scholarship Amount (converted)  ...  \
11                          0.0                             0.0  ...   
19                          0.0                             0.0  ...   
24                          0.0                             0.0  ...   
26                          0.0                             0.0  ...   
33                          0.0                             0.0  ...   

    Net Amount With VAT (converted)  Drop    Drop Reason  \
11                         11061.95     0  Not specified   
19                          7500.00     0  Not specified   
24                          7500.00     0  Not specified   
26                          4000.00     0  Not specified   
33                          5950.00     0  Not specified   

   how did you hear about us? Person Account: Gender Stage Duration  \
11                     google                   Male          117.0   
19                     google                   Male           11.0   
24                   switchup                   Male           83.0   
26               Social Media                 Female          441.0   
33                      other                 Female          335.0   

    Paid Deposit  Scholarship  Discount  Time Conversion - days  
11             0            0         0                      84  
19             0            0         0                     205  
24             0            0         0                      45  
26             0            0         0                      26  
33             0            0         0                     172  

[5 rows x 21 columns]

In [122]:
apps_clean[apps_clean['Close Date'] < apps_clean['Created Date']].head()

Empty DataFrame
Columns: [Bootcamp Course, Bootcamp Format, Bootcamp Year, Bootcamp Start Date, Campus - Timezone, Created Date, Close Date, Stage, Discount Amount (converted), Scholarship Amount (converted), Amount (converted), Net Amount With VAT (converted), Drop, Drop Reason, how did you hear about us?, Person Account: Gender, Stage Duration, Paid Deposit, Scholarship, Discount, Time Conversion - days]
Index: []

[0 rows x 21 columns]

In [123]:
apps_clean.tail()

Bootcamp Course Bootcamp Format  Bootcamp Year Bootcamp Start Date  \
28718              WD              FT           2020          2020-10-19   
28719              DA              FT           2021          2021-11-01   
28720              WD              FT           2020          2020-10-19   
28721              UX              FT           2020          2020-09-28   
28722              DA              FT           2020          2020-10-19   

      Campus - Timezone Created Date Close Date         Stage  \
28718               BER   2020-07-20 2020-08-14  Paid Deposit   
28719               AMS   2020-07-20 2020-10-27  Paid Deposit   
28720               BER   2020-07-20 2020-08-25  Paid Deposit   
28721               BCN   2020-07-21 2020-09-30   Closed lost   
28722               BER   2020-08-07 2020-09-30   Closed lost   

       Discount Amount (converted)  Scholarship Amount (converted)  ...  \
28718                        750.0                             0.0  ...   
28719                        800.0                             0.0  ...   
28720                          0.0                             0.0  ...   
28721                          0.0                             0.0  ...   
28722                       2099.0                             0.0  ...   

       Net Amount With VAT (converted)  Drop    Drop Reason  \
28718                           6750.0     0  Not specified   
28719                           7200.0     0  Not specified   
28720                           7500.0     0  Not specified   
28721                           7000.0     0  Not specified   
28722                           5401.0     0  Not specified   

      how did you hear about us? Person Account: Gender Stage Duration  \
28718                     google                 Female          123.0   
28719                      other                 Female           20.0   
28720                   Referral                   Male          106.0   
28721              IronhackEvent                 Female           81.0   
28722               Social Media                 Female          118.0   

       Paid Deposit  Scholarship  Discount  Time Conversion - days  
28718             1            0         1                      25  
28719             1            0         1                      99  
28720             1            0         0                      36  
28721             0            0         0                      71  
28722             0            0         1                      54  

[5 rows x 21 columns]

#### Add Column: Time between Created Date and Start Date (days)

In [124]:

apps_clean["Time between Created Date and Start Date - days"]= (apps_clean['Bootcamp Start Date'] - apps_clean['Created Date']).dt.days


#### Add Column: % Converstion Rate


In [125]:
bst_campus_course_format= apps_clean[apps_clean["Paid Deposit"]==1].groupby(["Campus - Timezone","Bootcamp Course","Bootcamp Format",
                    "Bootcamp Start Date"]).count()["Created Date"]


total_campus_course_format= apps_clean.groupby(["Campus - Timezone","Bootcamp Course","Bootcamp Format",
                    "Bootcamp Start Date"]).count()["Created Date"]


In [126]:
cvr_campus_course_format=round(bst_campus_course_format/total_campus_course_format,2)

In [127]:
cvr_campus_course_format.fillna(0, inplace = True) 

In [128]:
new_df=pd.DataFrame(cvr_campus_course_format)
new_df

Created Date
Campus - Timezone Bootcamp Course Bootcamp Format Bootcamp Start Date              
AMS               CY              FT              2021-07-06                   0.00
                  DA              FT              2019-09-23                   0.17
                                                  2020-02-17                   0.26
                                                  2020-04-05                   0.27
                                                  2020-06-15                   0.00
...                                                                             ...
SAO               WD              PT              2020-07-07                   0.33
                                                  2020-08-01                   0.04
                                                  2020-09-22                   0.00
                                                  2020-10-20                   0.45
                                                  2021-12-01                   0.13

[870 rows x 1 columns]

This column was not finally added as was gonna bring colliniarity with other columns

#### Add Column: % Discount

In [130]:
apps_clean["Discount(%)"]= 1- (round(apps_clean["Net Amount With VAT (converted)"]  / apps_clean["Amount (converted)"],2) )


#### Add Column: Creater Month 
Aggregate Creater Date by month

In [131]:
apps_clean["Creater Month"]=apps_clean['Created Date'].dt.month

#### Add Column: Creater Quarter 
Aggregate Creater Date by quarter

In [132]:
apps_clean["Creater Quarter"]=apps_clean['Created Date'].dt.quarter

#### Drop columns used to create the new ones and that now have no more use

In [133]:
apps_clean.columns

Index(['Bootcamp Course', 'Bootcamp Format', 'Bootcamp Year',
       'Bootcamp Start Date', 'Campus - Timezone', 'Created Date',
       'Close Date', 'Stage', 'Discount Amount (converted)',
       'Scholarship Amount (converted)', 'Amount (converted)',
       'Net Amount With VAT (converted)', 'Drop', 'Drop Reason',
       'how did you hear about us?', 'Person Account: Gender',
       'Stage Duration', 'Paid Deposit', 'Scholarship', 'Discount',
       'Time Conversion - days',
       'Time between Created Date and Start Date - days', 'Discount(%)',
       'Creater Month', 'Creater Quarter'],
      dtype='object')

In [134]:
apps_clean.drop(columns=["Bootcamp Start Date", "Created Date", "Close Date",
                        "Discount Amount (converted)", "Scholarship Amount (converted)",
                        "Amount (converted)", "Net Amount With VAT (converted)",
                        "Stage"], inplace=True)

In [158]:
apps_clean.columns

Index(['Bootcamp Course', 'Bootcamp Format', 'Bootcamp Year',
       'Campus - Timezone', 'Drop', 'Drop Reason',
       'how did you hear about us?', 'Person Account: Gender',
       'Stage Duration', 'Paid Deposit', 'Scholarship', 'Discount',
       'Time Conversion - days',
       'Time between Created Date and Start Date - days', 'Discount(%)',
       'Creater Month', 'Creater Quarter'],
      dtype='object')

In [161]:
apps_clean.shape

(28723, 17)

## CleanCSV 
Save CSV ready to model

In [155]:
apps_clean.to_csv("apps_allYears_clean_selCols_addCols.csv", index=False)

## Insights 

- We must start **unifying labels** from each column (example: Gender,How did you hear about us?,...
- **Wrong years such 1970 & 1900** are shown (actually it was data from 2016) 
- We should **work with data from campus that finally didn't open** (such us Munich, Bogota and RMT-Chicago) and find out if they were redirected to any ther campus. If not, we should study how "hot" those application are.
- We have **missing but also wrong dates** (close dates, close month and invoice dates) - not from any specific year, campus or bootcamps type. For example close dates are wrong as 20% of them are smaller then the created date plus 1% is missing.
- **Running campaings without selected start data makes it complicated** to later analyse in detail. Plus, we've seen that most of them don't convert. It might make more sense to put is as "ask for info" and also include a form of what type of info we would like to have.
- Person Account data such us Looking for a Job, Tech Background,Highest degree received,country, city should be well recollected during PIs. **Demographic data is the most missing data**.
- Paid Deposit Status is not being linked to stage.

- **We've been able to clean data from 12 columns out of 58 columns.
- **We'll be working with 17 variables from now on incl those we added (not norml cols).**
  Added columns: Paid Deposit(target), Time Conversion (days),Time between Created Date and Start Date (days), %Converstion rate, %Discount,Creater Month,Creater Quarter.

“In God we trust, all others must bring data" W. Edwards Deming